app/src/main/java/dadeindustries/game/gc/view/GalaxyView.java

```diff
@@ -174,150 +177,253 @@ else if (x > (GlobalGameData.galaxySizeX - NUM_SQUARES_IN_ROW)) {
+		paint.setColor(Color.WHITE);
+		// Put text in top left corner indicating the current turn number
+		canvas.drawText("Turn " + globalGameData.getTurn(),
+				viewPort.x + PADDING, viewPort.y + PADDING * 3, paint);
+		canvas.drawText("Credits " + globalGameData.getHumanPlayerCredits(),
+				viewPort.x + PADDING, viewPort.y + PADDING * 3 * 2, paint);
+		paint.setColor(savedColor);
+	}
 
-				int shipx = i;
-				int shipy = j;
+	/**
+	 * Draws the bitmap of a ship at a particular sector
+	 *
+	 * @param canvas
+	 * @param sector
+	 */
+	public void drawShip(Canvas canvas, Sector sector) {
 
+		int x, y;
+		int shipx = sector.getX();
+		int shipy = sector.getY();
 
-				if ((shipx >= viewPort.x)
-						&& (shipx <= viewPort.x + NUM_SQUARES_IN_ROW)
-						&& (shipy >= viewPort.y)) {
+		if ((shipx >= viewPort.x)
+				&& (shipx <= viewPort.x + NUM_SQUARES_IN_ROW)
+				&& (shipy >= viewPort.y)
+				&& globalGameData.getHumanPlayer().isVisible(sector) == true
+				) {
 
-					int x = (shipx - viewPort.x) * SQUARE_SIZE;
-					int y = (shipy - viewPort.y) * SQUARE_SIZE;
-					r.left = x;
-					r.top = y;
-					r.right = x + (SQUARE_SIZE / 2) - PADDING;
-					r.bottom = y + (SQUARE_SIZE / 2) - PADDING;
+			x = (shipx - viewPort.x) * SQUARE_SIZE;
+			y = (shipy - viewPort.y) * SQUARE_SIZE;
+			r.left = x;
+			r.top = y;
+			r.right = x + SQUARE_SIZE / 2;
+			r.bottom = y + SQUARE_SIZE / 2;
 
-					// TODO: Need to handle case where multiple factions are in the same system
+			// TODO: Need to handle case where multiple units in the same system
 
-					ArrayList<Spaceship> ships = sectors[i][j].getUnits();
-					for (Spaceship ship : ships) {
+			ArrayList<Spaceship> ships = sector.getUnits();
+			for (Spaceship ship : ships) {
 
-						switch (ship.getOwner().getIntelligence()) {
+				switch (ship.getOwner().getIntelligence()) {
 
-							case HUMAN:
-								canvas.drawBitmap(up, null, r, paint);
-								break;
+					case HUMAN:
+						getGlobalGameData().getHumanPlayer().discover(sector);
+						canvas.drawBitmap(up, null, r, paint);
+						break;
 
-							case ARTIFICIAL:
-								canvas.drawBitmap(mo, null, r, paint);
-								break;
+					case ARTIFICIAL:
+						canvas.drawBitmap(mo, null, r, paint);
+						break;
 
-							default:
-								// do nothing
-						}
+					default:
+						// do nothing
+				}
 
-						canvas.drawText(ship.getShipName(), x + PADDING, y + (PADDING * 3)
-								+ (SQUARE_SIZE / 2), paint);
-					}
+				paint.setColor(Color.WHITE);
+				canvas.drawText(ship.getShipName(), x + PADDING, y + (PADDING * 3)
+						+ (SQUARE_SIZE / 2), paint);
+			}
+		}
+	}
 
+	/**
+	 * This is triggered every time the screen refreshes/repaints
+	 *
+	 * @param canvas
+	 */
+	@Override
+	public void onDraw(Canvas canvas) {
+
+		/* Paint all sectors */
+		for (int i = viewPort.x; i < (viewPort.x + NUM_SQUARES_IN_ROW); i++) {
+			for (int j = viewPort.y; j < (viewPort.y + NUM_SQUARES_IN_COLUMN); j++) {
+
+				Sector sector = sectors[i][j];
```

Comment:
This line gives an ArrayOutOfBoundsException: length 10, index -5.

Reply:
This is because the screen scrolling is broken on your phone and takes the map out of bounds. Can you try this with the Galaxy Nexus version of the emulator in the meantime?

I can look into fixing the scrolling bug when I get home, which should also fix this problem.